crates/kitsune_p2p/kitsune_p2p/src/test_util/harness_actor.rs

```diff
@@ -0,0 +1,308 @@
+use super::*;
+
+ghost_actor::ghost_chan! {
+    /// The api for the test harness controller
+    pub chan HarnessControlApi<KitsuneP2pError> {
+        /// Create a new random space id
+        /// + join all existing harness agents to it
+        /// + all new harness agents will also join it
+        fn add_space() -> Arc<KitsuneSpace>;
+
+        /// Create a new agent configured to proxy for others.
+        fn add_proxy_agent(nick: String) -> (
+            Arc<KitsuneAgent>,
+            ghost_actor::GhostSender<KitsuneP2p>,
+        );
+
+        /// Create a new directly addressable agent that will
+        /// reject any proxy requests.
+        fn add_direct_agent(nick: String) -> (
+            Arc<KitsuneAgent>,
+            ghost_actor::GhostSender<KitsuneP2p>,
+        );
+
+        /// Create a new agent that will connect via proxy.
+        fn add_nat_agent(nick: String, proxy_url: url2::Url2) -> (
+            Arc<KitsuneAgent>,
+            ghost_actor::GhostSender<KitsuneP2p>,
+        );
+
+        /// Inject data for one specific agent to gossip to others
+        fn inject_gossip_data(agent: Arc<KitsuneAgent>, data: String) -> Arc<KitsuneOpHash>;
+
+        /// Dump all local gossip data from a specific agent
+        fn dump_local_gossip_data(agent: Arc<KitsuneAgent>) -> HashMap<Arc<KitsuneOpHash>, String>;
+    }
+}
+
+/// construct a test suite around a mem transport
+pub async fn spawn_test_harness_mem() -> Result<
+    (
+        ghost_actor::GhostSender<HarnessControlApi>,
+        HarnessEventChannel,
+    ),
+    KitsuneP2pError,
+> {
+    spawn_test_harness(TransportConfig::Mem {}).await
+}
+
+/// construct a test suite around a quic transport
+pub async fn spawn_test_harness_quic() -> Result<
+    (
+        ghost_actor::GhostSender<HarnessControlApi>,
+        HarnessEventChannel,
+    ),
+    KitsuneP2pError,
+> {
+    spawn_test_harness(TransportConfig::Quic {
+        bind_to: Some(url2::url2!("kitsune-quic://0.0.0.0:0")),
+        override_host: None,
+        override_port: None,
+    })
+    .await
+}
+
+/// construct a test suite around a sub transport config concept
+pub async fn spawn_test_harness(
+    sub_config: TransportConfig,
+) -> Result<
+    (
+        ghost_actor::GhostSender<HarnessControlApi>,
+        HarnessEventChannel,
+    ),
+    KitsuneP2pError,
+> {
+    init_tracing();
```

Comment:
If this function can be called from a `holochain` test then initing tracing here will break the `holochain` tracing

Reply:
yar, good point... not sure how to solve that...

Reply:
ok, pulled it out so you now have to manually init_tracing in every test... would be nice to get rid of that boilerplate somehow some day : )